From 48440dee469cde48ee92bb1ea53c5d8037d5586d Mon Sep 17 00:00:00 2001
From: Alf Henrik Sauge <alf.henrik.sauge@gmail.com>
Date: Fri, 14 Aug 2026 13:24:37 +0200
Subject: [PATCH] Fix src/git/Filter.cpp passing data into the shell without
 escaping it

Without escaping the data, remote shell execution can occur, as evidenced by the test case
---
 src/git/Filter.cpp                        |   8 +-
 1 file changed, 16 insertions(+), 1 deletion(-)
 create mode 100644 test/Filter.cpp
 create mode 100644 test/testRepos/FilterCommandInjection.zip

diff --git a/src/git/Filter.cpp b/src/git/Filter.cpp
index 91d0f1423..c07ca3a69 100644
--- a/src/git/Filter.cpp
+++ b/src/git/Filter.cpp
@@ -38,7 +38,13 @@ struct FilterInfo {
   QByteArray attributes;
 };
 
-QString quote(const QString &path) { return QString("\"%1\"").arg(path); }
+QString quote(const QString &path) {
+  QString escapedPath = path;
+  // Ensure that the path is properly escaped to avoid shell injection
+  // This is inspired by git's sq_quote_buf
+  escapedPath.replace("'", "'\\''");
+  return QString("'%1'").arg(escapedPath);
+}
 
 struct Stream {
   int init(git_filter *self, const git_filter_source *src, git_writestream *);
